udfs/tests/udf_test_utils.py

```diff
@@ -64,63 +122,205 @@ def extract_udf_name(udf_path):
+    """
+    For a given path to a UDF DDL file, parse the SQL and return the UDF
+    with the dataset name changed with an added suffix for testing. The suffix
+    value is defined in the global variable, DATASET_SUFFIX, and includes the
+    commit's SHORT_SHA to prevent different commits from interfering
+    with UDF builds.
 
-def replace_with_test_datasets(udf_path=None, project_id=None, udf_sql=None):
-    if udf_path:
-        with open(udf_path) as udf_file:
-            udf_sql = udf_file.read()
-    udf_length_before_replacement = len(udf_sql)
-    udf_sql = re.sub(
+    :param project_id: Project in which to create the UDF
+    :param udf_sql: SQL DDL of the UDF
+    :return: Same SQL DDL as input udf_sql but replaced with testing dataset
+    """
+    udf_sql_with_test_dataset = re.sub(
         r'(\w+\.)?(?P<bq_dataset>\w+)(?P<udf_name>\.\w+)\(',
-        f'`{project_id}.\\g<bq_dataset>_test_{os.getenv("SHORT_SHA")}\\g<udf_name>`(',
+        f'`{project_id}.\\g<bq_dataset>{DATASET_SUFFIX}\\g<udf_name>`(',
         udf_sql)
-    if udf_length_before_replacement == len(udf_sql):
+    if udf_sql_with_test_dataset == udf_sql:
+        # Replacement failed, return None to prevent overwriting prod dataset
         return None
     else:
-        return udf_sql
+        return udf_sql_with_test_dataset
 
 
-def get_target_bq_dataset(udf_path):
-    parent_dir_name = Path(udf_path).parent.name
-    return f'{BIGQUERY_TEST_DATASET_MAPPINGS.get(parent_dir_name)}_{os.getenv("SHORT_SHA")}'
+def get_test_bq_dataset(udf_path):
+    """
+    For a given path to a UDF DDL file, return the BigQuery dataset name in
+    which to test the UDF. The test dataset name is the same as production
+    but with the suffix, _test_$SHORT_SHA, appended. The $SHORT_SHA value comes
+    from the commit which triggered the build.
+    :param udf_path: Path to the UDF DDL .sql file
+    :return: Name of test dataset
+    """
+    udf_parent_dir_name = Path(udf_path).parent.name
+    if os.getenv('SHORT_SHA') is not None:
+        bq_dataset = get_dir_to_dataset_mappings().get(udf_parent_dir_name)
+        return f'{bq_dataset}{DATASET_SUFFIX}'
+    else:
+        return None
 
 
 def delete_datasets(client):
-    for dataset in BIGQUERY_TEST_DATASET_MAPPINGS.values():
-        dataset = f'{dataset}_{os.getenv("SHORT_SHA")}'
+    for dataset in get_dir_to_dataset_mappings().values():
+        dataset = f'{dataset}{DATASET_SUFFIX}'
         client.delete_dataset(dataset, delete_contents=True, not_found_ok=True)
 
 
-def create_datasets(client):
-    for dataset in BIGQUERY_TEST_DATASET_MAPPINGS.values():
-        dataset = f'{dataset}_{os.getenv("SHORT_SHA")}'
+def create_datasets(client, dataset_suffix=None):
+    for dataset in get_dir_to_dataset_mappings().values():
+        if dataset_suffix:
+            dataset = f'{dataset}{dataset_suffix}'
         client.create_dataset(dataset, exists_ok=True)
 
 
+def generate_js_libs_package_json():
```

Comment:
this function is extremely hard to read / follow.
I'd recommend editing this as a dict object before writing to file.
I'm not sure I quite got this right but it will illustrate how to clean this up.
```python3
package_json_dict = 
{
    "name": "js-bq-libs",
    "version": "1.0.0",
    "scripts": {
        "build-all-libs": "concurrently \\"npm:webpack-*\\""
    },
    "dependencies": {
        f'{lib_name}-v{version}" ': f'npm:{lib_name}@^{version}'
        for lib_name in js_libs_dict
        for version in js_libs_dict.get(lib_name).get('versions')
    },
    "devDependencies": {
        "webpack": "^5.3.1",
        "webpack-cli": "^4.1.0",
        "concurrently": "^5.3.0"
        }'
}

package_json['scripts'].update({
    f'webpack-{lib_name}-v{version}':
    f'webpack --config {lib_name}-v{version}-webpack.config.js'
    for lib_name in js_libs_dict
    for version in js_libs_dict.get(lib_name).get('versions')                
})

with open('./package.json', 'w') as js_libs_package_json:
    json.dump(package_json, js_libs_package_json)
```

Alternatively you could make a template string with sections.

Reply:
fantastic recommendation! much cleaner